File: app/src/main/java/md/fusionworks/lifehack/ui/exchange_rates/fragment/BranchMapFragment.java
package md.fusionworks.lifehack.ui.exchange_rates.fragment;

import android.location.Location;
import android.os.Bundle;
import android.support.annotation.Nullable;
import android.support.v4.app.Fragment;
import android.support.v4.content.ContextCompat;
import android.view.LayoutInflater;
import android.view.MotionEvent;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import butterknife.Bind;
import com.badoo.mobile.util.WeakHandler;
import com.directions.route.AbstractRouting;
import com.directions.route.Route;
import com.directions.route.RouteException;
import com.directions.route.Routing;
import com.directions.route.RoutingListener;
import com.google.android.gms.common.api.GoogleApiClient;
import com.google.android.gms.location.LocationServices;
import com.google.android.gms.maps.GoogleMap;
import com.google.android.gms.maps.MapView;
import com.google.android.gms.maps.MapsInitializer;
import com.google.android.gms.maps.model.LatLng;
import com.google.android.gms.maps.model.Marker;
import com.google.android.gms.maps.model.Polyline;
import com.google.android.gms.maps.model.PolylineOptions;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import md.fusionworks.lifehack.R;
import md.fusionworks.lifehack.ui.exchange_rates.event.ScrollToMapEvent;
import md.fusionworks.lifehack.ui.exchange_rates.event.ShowBranchMapInfoWindowEvent;
import md.fusionworks.lifehack.ui.exchange_rates.event.ShowRouteOnMapEvent;
import md.fusionworks.lifehack.ui.base.view.BaseFragment;
import md.fusionworks.lifehack.ui.exchange_rates.model.BranchModel;
import md.fusionworks.lifehack.util.Constant;
import md.fusionworks.lifehack.util.DialogUtil;
import md.fusionworks.lifehack.util.MapUtil;

/**
 * A simple {@link Fragment} subclass.
 */
public class BranchMapFragment extends BaseFragment {

  @Bind(R.id.rootLayout) FrameLayout rootLayout;
  @Bind(R.id.branchMap) MapView mapView;
  @Bind(R.id.imageOverMap) ImageView imageOverMap;

  private GoogleMap map;
  private Map<Marker, BranchModel> branchMap = new HashMap<>();
  private GoogleApiClient googleApiClient;
  private Location myLastLocation;
  private ArrayList<Polyline> routePolylines;
  private List<BranchModel> branchModelList;
  private WeakHandler weakHandler;

  public BranchMapFragment() {
  }

  public static BranchMapFragment newInstance(List<BranchModel> branchModelList) {
    BranchMapFragment branchMapFragment = new BranchMapFragment();
    Bundle bundle = new Bundle();
    bundle.putParcelableArrayList(Constant.EXTRA_PARAM_BRANCH_LIST,
        new ArrayList<>(branchModelList));
    branchMapFragment.setArguments(bundle);
    return branchMapFragment;
  }

  @Override public void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    branchModelList = getArguments().getParcelableArrayList(Constant.EXTRA_PARAM_BRANCH_LIST);
    routePolylines = new ArrayList<>();
    weakHandler = new WeakHandler();
    buildGoogleApiClient();
  }

  @Override public void onStart() {
    super.onStart();
    googleApiClient.connect();
  }

  @Override public void onStop() {
    super.onStop();
    if (googleApiClient.isConnected()) {
      googleApiClient.disconnect();
    }
  }

  @Override public void onResume() {
    super.onResume();
    mapView.onResume();
  }

  @Override public void onDestroy() {
    weakHandler.removeCallbacksAndMessages(null);
    if (map != null) {
      map.setMyLocationEnabled(false);
      map.clear();
    }
    mapView.onDestroy();
    super.onDestroy();
  }

  @Override public void onLowMemory() {
    super.onLowMemory();
    mapView.onLowMemory();
  }

  @Override public View onCreateView(LayoutInflater inflater, ViewGroup container,
      Bundle savedInstanceState) {
    View view = inflateAndBindViews(inflater, R.layout.fragment_branch_map, container);
    initializeMap(savedInstanceState);
    return view;
  }

  @Override protected void listenForEvents() {
    super.listenForEvents();
    getRxBus().event(ShowRouteOnMapEvent.class)
        .compose(this.bindToLifecycle())
        .subscribe(showRouteOnMapEvent -> {
          showRouteOnMap(showRouteOnMapEvent.getBranchModel());
        });

    getRxBus().event(ShowBranchMapInfoWindowEvent.class)
        .compose(this.bindToLifecycle())
        .subscribe(showBranchMapInfoWindowEvent -> {
          smoothShowInfoWindow(showBranchMapInfoWindowEvent.getBranchModel());
        });
  }

  private void initializeMap(Bundle savedInstanceState) {
    mapView.onCreate(savedInstanceState);
    enableMapGesturesInScrollView();
    mapView.getMapAsync(googleMap -> {
      map = googleMap;
      map.getUiSettings().setMyLocationButtonEnabled(true);
      map.getUiSettings().setZoomControlsEnabled(true);
      map.getUiSettings().setCompassEnabled(true);
      map.setMyLocationEnabled(true);
      MapsInitializer.initialize(getActivity());
      populateBranchesMap(branchModelList);
      weakHandler.postDelayed(() -> getRxBus().postIfHasObservers(new ScrollToMapEvent()), 500);
      map.setOnMarkerClickListener(marker -> {
        smoothShowInfoWindow(branchMap.get(marker));
        return true;
      });
    });
  }

  private void enableMapGesturesInScrollView() {
    imageOverMap.setOnTouchListener((v, event) -> {
      int action = event.getAction();
      switch (action) {
        case MotionEvent.ACTION_DOWN:
          rootLayout.requestDisallowInterceptTouchEvent(true);
          return false;
        case MotionEvent.ACTION_UP:
          rootLayout.requestDisallowInterceptTouchEvent(false);
          return true;
        case MotionEvent.ACTION_MOVE:
          rootLayout.requestDisallowInterceptTouchEvent(true);
          return false;
        default:
          return true;
      }
    });
  }

  private void pinHomeMarker(Location location) {
    if (location != null) {
      MapUtil.createMarker(map, location.getLatitude(), location.getLongitude(),
          R.drawable.home_pin_icon);
      MapUtil.goToPosition(map, location.getLatitude(), location.getLongitude(), false,
          Constant.CAMERA_ZOOM);
    }
  }

  private void populateBranchesMap(List<BranchModel> branchList) {
    map.clear();
    pinHomeMarker(myLastLocation);
    for (BranchModel branch : branchList) {
      Marker marker = MapUtil.createMarker(map, branch.getAddress().getLocation().getLat(),
          branch.getAddress().getLocation().getLng(), R.drawable.exchange_pin_icon);
      branchMap.put(marker, branch);
    }
  }

  private void smoothShowInfoWindow(BranchModel branch) {
    weakHandler.postDelayed(() -> DialogUtil.showBranchMapInfoWindow(getActivity(), branch), 100);
    weakHandler.postDelayed(
        () -> MapUtil.goToPosition(map, branch.getAddress().getLocation().getLat(),
            branch.getAddress().getLocation().getLng(), false), 200);
  }

  protected synchronized void buildGoogleApiClient() {
    googleApiClient = new GoogleApiClient.Builder(getActivity()).addConnectionCallbacks(
        new GoogleApiClient.ConnectionCallbacks() {
          @Override public void onConnected(@Nullable Bundle bundle) {
            myLastLocation = LocationServices.FusedLocationApi.getLastLocation(googleApiClient);
            pinHomeMarker(myLastLocation);
          }

          @Override public void onConnectionSuspended(int i) {
            googleApiClient.connect();
          }
        }).addOnConnectionFailedListener(connectionResult -> {

    }).addApi(LocationServices.API).build();
  }

  private void showRouteOnMap(BranchModel branch) {
    if (myLastLocation != null) {
      showLoadingDialog();
      Routing routing = new Routing.Builder().travelMode(AbstractRouting.TravelMode.DRIVING)
          .withListener(new RoutingListener() {
            @Override public void onRoutingFailure(RouteException e) {
              hideLoadingDialog();
              showNotificationToast(Constant.NOTIFICATION_TOAST_ERROR,
                  getString(R.string.error_something_gone_wrong));
            }

            @Override public void onRoutingStart() {

            }

            @Override public void onRoutingSuccess(ArrayList<Route> arrayList, int j) {
              hideLoadingDialog();
              MapUtil.goToPosition(map, branch.getAddress().getLocation().getLat(),
                  branch.getAddress().getLocation().getLng(), false);

              if (routePolylines.size() > 0) {
                for (Polyline poly : routePolylines) {
                  poly.remove();
                }
              }
              routePolylines = new ArrayList<>(arrayList.size());

              for (int i = 0; i < arrayList.size(); i++) {
                PolylineOptions polyOptions = new PolylineOptions();
                polyOptions.color(ContextCompat.getColor(getActivity(), R.color.mainColorBlue));
                polyOptions.width(10 + i * 3);
                polyOptions.addAll(arrayList.get(i).getPoints());
                Polyline polyline = map.addPolyline(polyOptions);
                routePolylines.add(polyline);
              }
            }

            @Override public void onRoutingCancelled() {

            }
          })
          .alternativeRoutes(false)
          .waypoints(new LatLng(myLastLocation.getLatitude(), myLastLocation.getLongitude()),
              new LatLng(branch.getAddress().getLocation().getLat(),
                  branch.getAddress().getLocation().getLng()))
          .build();
      routing.execute();
    } else {
      showNotificationToast(Constant.NOTIFICATION_TOAST_ERROR,
          getString(R.string.error_something_gone_wrong));
    }
  }
}

File: app/src/main/java/md/fusionworks/lifehack/exchange_rates/fragment/BranchMapFragment.kt
package md.fusionworks.lifehack.exchange_rates.fragment

import android.location.Location
import android.os.Bundle
import android.support.v4.content.ContextCompat
import android.view.LayoutInflater
import android.view.MotionEvent
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ImageView
import com.badoo.mobile.util.WeakHandler
import com.directions.route.*
import com.google.android.gms.common.api.GoogleApiClient
import com.google.android.gms.location.LocationServices
import com.google.android.gms.maps.GoogleMap
import com.google.android.gms.maps.MapView
import com.google.android.gms.maps.MapsInitializer
import com.google.android.gms.maps.model.LatLng
import com.google.android.gms.maps.model.Marker
import com.google.android.gms.maps.model.Polyline
import com.google.android.gms.maps.model.PolylineOptions
import md.fusionworks.lifehack.R
import md.fusionworks.lifehack.view.fragment.BaseFragment
import md.fusionworks.lifehack.exchange_rates.event.ScrollToMapEvent
import md.fusionworks.lifehack.exchange_rates.event.ShowBranchMapInfoWindowEvent
import md.fusionworks.lifehack.exchange_rates.event.ShowRouteOnMapEvent
import md.fusionworks.lifehack.exchange_rates.model.BranchModel
import md.fusionworks.lifehack.util.Constant
import md.fusionworks.lifehack.util.DialogUtil
import md.fusionworks.lifehack.helper.MapHelper
import md.fusionworks.lifehack.rx.RxBus
import java.util.*

/**
 * A simple [Fragment] subclass.
 */
class BranchMapFragment : BaseFragment() {

  private lateinit var branchMapView: MapView
  private lateinit var imageOverMap: ImageView
  private lateinit var rootLayout: FrameLayout


  private lateinit var map: GoogleMap
  private val branchMap = HashMap<Marker, BranchModel>()
  private lateinit var googleApiClient: GoogleApiClient
  private var myLastLocation: Location? = null
  private lateinit var routePolylines: ArrayList<Polyline>
  private lateinit var branchModelList: List<BranchModel>
  private lateinit var weakHandler: WeakHandler

  override fun onCreate(savedInstanceState: Bundle?) {
    super.onCreate(savedInstanceState)
    branchModelList = arguments.getParcelableArrayList<BranchModel>(
        Constant.EXTRA_PARAM_BRANCH_LIST)
    routePolylines = arrayListOf<Polyline>()
    weakHandler = WeakHandler()

    buildGoogleApiClient()
  }

  override fun onStart() {
    super.onStart()
    googleApiClient.connect()
  }

  override fun onStop() {
    super.onStop()
    if (googleApiClient.isConnected) {
      googleApiClient.disconnect()
    }
  }

  override fun onResume() {
    super.onResume()
    branchMapView.onResume()
  }

  override fun onDestroy() {
    weakHandler.removeCallbacksAndMessages(null)
    if (map != null) {
      map.isMyLocationEnabled = false
      map.clear()
    }
    branchMapView.onDestroy()
    super.onDestroy()
  }

  override fun onLowMemory() {
    super.onLowMemory()
    branchMapView.onLowMemory()
  }

  override fun onCreateView(inflater: LayoutInflater?, container: ViewGroup?,
      savedInstanceState: Bundle?): View? {
    val view = inflater?.inflate(R.layout.fragment_branch_map, container, false)
    branchMapView = view?.findViewById(R.id.branchMapView) as MapView
    imageOverMap = view?.findViewById(R.id.imageOverMap) as ImageView
    rootLayout = view?.findViewById(R.id.rootLayout) as FrameLayout
    initializeMap(savedInstanceState)
    return view
  }

  override fun listenForEvents() {
    super.listenForEvents()
    RxBus.event(ShowRouteOnMapEvent::class.java).compose(
        this.bindToLifecycle<ShowRouteOnMapEvent>()).subscribe { showRouteOnMapEvent ->
      showRouteOnMap(showRouteOnMapEvent.branchModel)
    }

    RxBus.event(ShowBranchMapInfoWindowEvent::class.java).compose(
        this.bindToLifecycle<ShowBranchMapInfoWindowEvent>()).subscribe { showBranchMapInfoWindowEvent ->
      smoothShowInfoWindow(showBranchMapInfoWindowEvent.branchModel)
    }
  }

  private fun initializeMap(savedInstanceState: Bundle?) {
    branchMapView.onCreate(savedInstanceState)
    enableMapGesturesInScrollView()
    branchMapView.getMapAsync { googleMap ->
      map = googleMap
      map.uiSettings.isMyLocationButtonEnabled = true
      map.uiSettings.isZoomControlsEnabled = true
      map.uiSettings.isCompassEnabled = true
      map.isMyLocationEnabled = true
      MapsInitializer.initialize(activity)
      populateBranchesMap(branchModelList)
      weakHandler.postDelayed({ RxBus.postIfHasObservers(ScrollToMapEvent()) }, 500)
      map.setOnMarkerClickListener { marker ->
        smoothShowInfoWindow(branchMap[marker])
        true
      }
    }
  }

  private fun enableMapGesturesInScrollView() {
    imageOverMap.setOnTouchListener { v, event ->
      val action = event.action
      when (action) {
        MotionEvent.ACTION_DOWN -> {
          rootLayout.requestDisallowInterceptTouchEvent(true)
          false
        }
        MotionEvent.ACTION_UP -> {
          rootLayout.requestDisallowInterceptTouchEvent(false)
          true
        }
        MotionEvent.ACTION_MOVE -> {
          rootLayout.requestDisallowInterceptTouchEvent(true)
          false
        }
        else -> true
      }
    }
  }

  private fun pinHomeMarker(location: Location?) {
    if (location != null) {
      MapHelper.createMarker(map, location.latitude, location.longitude,
          R.drawable.home_pin_icon)
      MapHelper.goToPosition(map, location.latitude, location.longitude, false,
          Constant.CAMERA_ZOOM)
    }
  }

  private fun populateBranchesMap(branchList: List<BranchModel>?) {
    map.clear()
    pinHomeMarker(myLastLocation)
    if (branchList != null) for (branch in branchList) {
      val marker = MapHelper.createMarker(map, branch.address.location.lat,
          branch.address.location.lng, R.drawable.exchange_pin_icon)
      branchMap.put(marker, branch)
    }
  }

  private fun smoothShowInfoWindow(branch: BranchModel?) {
    weakHandler.postDelayed({ DialogUtil.showBranchMapInfoWindow(activity, branch) }, 100)
    weakHandler.postDelayed(
        {
          MapHelper.goToPosition(map, branch?.address?.location?.lat,
              branch?.address?.location?.lng, false)
        }, 200)
  }

  @Synchronized protected fun buildGoogleApiClient() {
    googleApiClient = GoogleApiClient.Builder(activity).addConnectionCallbacks(
        object : GoogleApiClient.ConnectionCallbacks {
          override fun onConnected(bundle: Bundle?) {
            myLastLocation = LocationServices.FusedLocationApi.getLastLocation(googleApiClient)
            pinHomeMarker(myLastLocation)
          }

          override fun onConnectionSuspended(i: Int) {
            googleApiClient.connect()
          }
        }).addOnConnectionFailedListener { connectionResult ->

    }.addApi(LocationServices.API).build()
  }

  private fun showRouteOnMap(branch: BranchModel?) {
    if (myLastLocation != null) {
      showLoadingDialog()
      val routing = Routing.Builder().travelMode(AbstractRouting.TravelMode.DRIVING).withListener(
          object : RoutingListener {
            override fun onRoutingFailure(e: RouteException) {
              hideLoadingDialog()
              showNotificationToast(Constant.NOTIFICATION_TOAST_ERROR,
                  getString(R.string.error_something_gone_wrong))
            }

            override fun onRoutingStart() {

            }

            override fun onRoutingSuccess(arrayList: ArrayList<Route>, j: Int) {
              hideLoadingDialog()
              MapHelper.goToPosition(map, branch?.address?.location?.lat,
                  branch?.address?.location?.lng, false)

              if (routePolylines.size > 0) {
                for (poly in routePolylines) {
                  poly.remove()
                }
              }
              routePolylines = ArrayList<Polyline>(arrayList.size)

              for (i in arrayList.indices) {
                val polyOptions = PolylineOptions()
                polyOptions.color(ContextCompat.getColor(activity, R.color.mainColorBlue))
                polyOptions.width((10 + i * 3).toFloat())
                polyOptions.addAll(arrayList[i].points)
                val polyline = map.addPolyline(polyOptions)
                routePolylines.add(polyline)
              }
            }

            override fun onRoutingCancelled() {

            }
          }).alternativeRoutes(false).waypoints(
          LatLng(myLastLocation!!.latitude, myLastLocation!!.longitude),
          LatLng(branch?.address?.location?.lat ?: 0.toDouble(),
              branch?.address?.location?.lng ?: 0.toDouble())).build()
      routing.execute()
    } else {
      showNotificationToast(Constant.NOTIFICATION_TOAST_ERROR,
          getString(R.string.error_something_gone_wrong))
    }
  }

  companion object {

    fun newInstance(branchModelList: List<BranchModel>?): BranchMapFragment {
      val branchMapFragment = BranchMapFragment()
      val bundle = Bundle()
      bundle.putParcelableArrayList(Constant.EXTRA_PARAM_BRANCH_LIST,
          ArrayList(branchModelList))
      branchMapFragment.arguments = bundle
      return branchMapFragment
    }
  }
}
